lib/spack/spack/compilers/__init__.py

```diff
@@ -611,17 +659,44 @@ def _default(cmp_id, paths):
         return [compiler]
 
     for compiler_id, by_compiler_id in compilers_d.items():
-        _, selected_name_variation = max(
-            (len(by_compiler_id[variation]), variation)
-            for variation in by_compiler_id
-        )
+        def sort_tuple(variation):
+            modules_for_paths = set()
+            paths = by_compiler_id[variation]
+            for p in paths.values():
+                modules_for_paths |= module_association.get(
+                    os.path.dirname(p), set())
+
+            positive_attributes = [
```

Comment:
This seems scary but I like it and it's pretty clear 😄

Will this select out the `intel` module for `icc` and the `gcc` module for `gcc` as I suggested above?

Reply:
This will select out the paths that map to a module that has nothing to do with the compiler.

For example, in LLNL's TCE there are symlinks of many of the compilers to `usr/tce/bin/name-version`, and a module called `StdEnv` that loads that path. We want to prefer the `usr/tce/packages/name/name-version/bin` compilers that are associated with modules `name/version`.

As I mentioned above, I think by this point the `intel` module already will not be associated with `gcc`.

Reply:
@becker33: we talked about making a module ineligible to be a `gcc` module if it loaded `gcc` and any other compiler -- basically make it so that if the `intel` module puts `icc` and `gcc` in the path, we *don't* make it eligible to be included in `compilers.yaml` for gcc.

Did that make it in here?  I'm not seeing which condition that corresponds to.